programs/drift/src/controller/orders.rs

```diff
@@ -668,6 +668,153 @@ pub fn cancel_order(
     Ok(())
 }
 
+pub enum ModifyOrderId {
+    UserOrderId(u8),
+    OrderId(u32),
+}
+
+pub fn modify_order(
+    order_id: ModifyOrderId,
+    modify_order_params: ModifyOrderParams,
+    user_loader: &AccountLoader<User>,
+    state: &State,
+    perp_market_map: &PerpMarketMap,
+    spot_market_map: &SpotMarketMap,
+    oracle_map: &mut OracleMap,
+    clock: &Clock,
+) -> DriftResult {
+    let user_key = user_loader.key();
+    let mut user = load_mut!(user_loader)?;
+    let order_index = match order_id {
+        ModifyOrderId::UserOrderId(user_order_id) => user
+            .get_order_index_by_user_order_id(user_order_id)
+            .map_err(|e| {
+                msg!("User order id {} not found", user_order_id);
+                e
+            })?,
+        ModifyOrderId::OrderId(order_id) => user.get_order_index(order_id).map_err(|e| {
+            msg!("Order id {} not found", order_id);
+            e
+        })?,
+    };
+    let existing_order = user.orders[order_index];
+
+    cancel_order(
+        order_index,
+        &mut user,
+        &user_key,
+        perp_market_map,
+        spot_market_map,
+        oracle_map,
+        clock.unix_timestamp,
+        clock.slot,
+        OrderActionExplanation::None,
+        None,
+        0,
+        false,
+    )?;
+
+    user.update_last_active_slot(clock.slot);
+
+    drop(user);
+
+    let order_params =
+        merge_modify_order_params_with_existing_order(&existing_order, &modify_order_params);
+
+    if order_params.market_type == MarketType::Perp {
+        place_perp_order(
+            state,
+            user_loader,
+            perp_market_map,
+            spot_market_map,
+            oracle_map,
+            clock,
+            order_params,
+        )?;
+    } else {
+        place_spot_order(
+            state,
+            user_loader,
+            perp_market_map,
+            spot_market_map,
+            oracle_map,
+            clock,
+            order_params,
+        )?;
+    }
+
+    Ok(())
+}
+
+fn merge_modify_order_params_with_existing_order(
+    existing_order: &Order,
+    modify_order_params: &ModifyOrderParams,
+) -> OrderParams {
+    let order_type = existing_order.order_type;
+    let market_type = existing_order.market_type;
+    let direction = modify_order_params
+        .direction
+        .unwrap_or(existing_order.direction);
+    let user_order_id = existing_order.user_order_id;
+    let base_asset_amount = modify_order_params
+        .base_asset_amount
+        .unwrap_or(existing_order.base_asset_amount);
```

Comment:
should this enforce that a partially filled order (base_asset_amount_filled!=0) must have the base asset amount replaced? just wondering if this is going to be super confusing state